### PR TITLE
[GlobalIsel] Combine extract vector element

### PR DESCRIPTION
**Pull request by tschuett**
look through shuffle vectors

### CODE REVIEW

File: llvm/lib/CodeGen/GlobalISel/CombinerHelperVectorOps.cpp [hunk 1]
```diff
@@ -325,6 +325,113 @@ bool CombinerHelper::matchExtractVectorElementWithBuildVectorTrunc(
...
+  // We try to get the value of the Index register.
+  std::optional<ValueAndVReg> MaybeIndex =
+      getIConstantVRegValWithLookThrough(Extract->getIndexReg(), MRI);
+  if (!MaybeIndex)
+    return false;
+
+  GShuffleVector *Shuffle =
+      cast<GShuffleVector>(getDefIgnoringCopies(Extract->getVectorReg(), MRI));
+
+  llvm::ArrayRef<int> Mask = Shuffle->getMask();
```
Comment (arsenm) [inline llvm/lib/CodeGen/GlobalISel/CombinerHelperVectorOps.cpp:381]: Don't need llvm::

File: llvm/lib/CodeGen/GlobalISel/CombinerHelperVectorOps.cpp [hunk 2]
```diff
@@ -325,6 +325,113 @@ bool CombinerHelper::matchExtractVectorElementWithBuildVectorTrunc(
...
+  //
+
+  // We try to get the value of the Index register.
+  std::optional<ValueAndVReg> MaybeIndex =
+      getIConstantVRegValWithLookThrough(Extract->getIndexReg(), MRI);
+  if (!MaybeIndex)
+    return false;
+
+  GShuffleVector *Shuffle =
+      cast<GShuffleVector>(getDefIgnoringCopies(Extract->getVectorReg(), MRI));
```
Comment (arsenm) [inline llvm/lib/CodeGen/GlobalISel/CombinerHelperVectorOps.cpp:379]: It's unfortunate you need to re-find the def instruction when it was already matched

File: llvm/test/CodeGen/AArch64/GlobalISel/combine-extract-vec-elt.mir [hunk 3]
```diff
@@ -571,3 +571,84 @@ body:             |
     RET_ReallyLR implicit $x0
 ...
 ---
+name:            extract_from_build_vector_shuffle_vector_undef
+alignment:       4
+liveins:
+  - { reg: '$x0' }
+  - { reg: '$x1' }
+frameInfo:
+  maxAlignment:    1
```
Comment (arsenm) [inline llvm/test/CodeGen/AArch64/GlobalISel/combine-extract-vec-elt.mir:580]: Don't need any of this

File: llvm/include/llvm/CodeGen/GlobalISel/GenericMachineInstrs.h [hunk 4]
```diff
@@ -294,6 +294,18 @@ class GBuildVectorTrunc : public GMergeLikeInstr {
   }
 };
 
+/// Respresents a G_SHUFFLE_VECTOR.
+class GShuffleVector : public GenericMachineInstr {
+public:
+  Register getSrc1Reg() const { return getOperand(1).getReg(); }
+  Register getSrc2Reg() const { return getOperand(2).getReg(); }
+  llvm::ArrayRef<int> getMask() const { return getOperand(3).getShuffleMask(); }
```
Comment (arsenm) [inline llvm/include/llvm/CodeGen/GlobalISel/GenericMachineInstrs.h:302]: Don't need llvm::

File: llvm/lib/CodeGen/GlobalISel/CombinerHelperVectorOps.cpp (hunk 1 above)
Comment (tschuett) [reply llvm/lib/CodeGen/GlobalISel/CombinerHelperVectorOps.cpp]: Done.

File: llvm/test/CodeGen/AArch64/GlobalISel/combine-extract-vec-elt.mir (hunk 3 above)
Comment (tschuett) [reply llvm/test/CodeGen/AArch64/GlobalISel/combine-extract-vec-elt.mir]: Done.

File: llvm/include/llvm/CodeGen/GlobalISel/GenericMachineInstrs.h (hunk 4 above)
Comment (tschuett) [reply llvm/include/llvm/CodeGen/GlobalISel/GenericMachineInstrs.h]: Done.

File: llvm/lib/CodeGen/GlobalISel/CombinerHelperVectorOps.cpp (hunk 2 above)
Comment (tschuett) [reply llvm/lib/CodeGen/GlobalISel/CombinerHelperVectorOps.cpp:379]: Yeah, but the cast and get def is safer than I KNOW WHAT I AM DOING.

File: llvm/include/llvm/CodeGen/GlobalISel/GenericMachineInstrs.h [hunk 5]
```diff
@@ -294,6 +294,18 @@ class GBuildVectorTrunc : public GMergeLikeInstr {
   }
 };
 
+/// Respresents a G_SHUFFLE_VECTOR.
```
Comment (madhur13490) [inline llvm/include/llvm/CodeGen/GlobalISel/GenericMachineInstrs.h:297]: Typo: Respresents -> Represents.

**Review by aemerson — APPROVED** [review]

**Pull request merged**